### PR TITLE
feature/TSP-5553_UpdateCreateFormControlMethod

### PR DESCRIPTION
**Pull request by starksweeneym**
# Updates
- TSP-5553 As a sdk user, I can create a form control on a target ref

### Important Notes
- Updated structs/models. 
- Updated CreateControlOnRef method. 
- Added validation.
- Updated Tests

### CODE REVIEW

File: pkg/domain/form_control_ref.go [hunk 1]
```diff
@@ -1,26 +1,80 @@
...
+	o.Description = formControl.Description
+	o.Name = formControl.Name
+	o.Key = formControl.ControlJSON.Key
+	o.Value = value
+	o.FormControl = formControl
+	o.FormControlId = formControl.Id
+
+	err = validate.Struct(o)
+	if err != nil {
+		return errors.New(errInvalidFormControlRef)
```
Comment (starkmichelsc) [inline pkg/domain/form_control_ref.go]: log and return actual error

File: pkg/domain/form_control_ref.go [hunk 2]
```diff
@@ -1,26 +1,80 @@
...
+		return errors.New(errNoValueProvided)
+	}
+
+	return nil
+}
+
+func (o *FormControlRef) BuildFormControlRefForCreate(formControl FormControl, ref string, value string) error {
+	err := formControl.Validate()
+	if err != nil {
+		return errors.New(fmt.Sprintf(errInvalidFormControl, formControl.Name))
```
Comment (starkmichelsc) [inline pkg/domain/form_control_ref.go]: log and return actual error

File: pkg/domain/form_control_ref.go [hunk 3]
```diff
@@ -1,26 +1,80 @@
...
+	}
+	err = json.Unmarshal([]byte(formControl.Control), &formControl.ControlJSON)
+	if err != nil {
```
Comment (starkmichelsc) [inline pkg/domain/form_control_ref.go:57]: log and return actual error

File: pkg/starkapi/forms_api.go [hunk 4]
```diff
@@ -80,30 +86,53 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+	url := fmt.Sprintf("%s/%s%s", formsApi.baseUrl(), ref, formsApi.controlsPrefix())
+
+	body, err := json.Marshal(controlRef)
+	if err != nil {
+		logger.Error(err)
+		return controlRef, err
 	}
 
 	resp, err := formsApi.client.post(url, body)
 	if err != nil {
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go:141]: log error

File: pkg/starkapi/forms_api.go [hunk 5]
```diff
@@ -80,30 +86,53 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
-		return control, err
+		return controlRef, err
 	}
 
 	if len(resp) > 0 {
-		err = json.Unmarshal(resp, &control)
+		err = json.Unmarshal(resp, &controlRef)
 		if err != nil {
-			return control, err
+			return controlRef, err
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go:150]: log error

File: pkg/starkapi/forms_api.go [hunk 6]
```diff
@@ -80,30 +86,53 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
 	return control, nil
 }
 
-func (formsApi *FormsApi) CreateControlOnRef(control domain.FormControl) (domain.FormControl, error) {
-	err := control.Validate()
+func (formsApi *FormsApi) CreateControlOnRef(formControlName string, ref string, value string) (domain.FormControlRef, error) {
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go:92]: Refactor to take in a formControl struct vs just the name. https://en.wikipedia.org/wiki/Single-responsibility_principle. the current method has multiple responsibilities. 1) create for control ref.  2) look up a formControl. in order to use this method a user would have to look up a formControl or a list of form controls to get a valid name

File: pkg/domain/form_control_ref.go [hunk 7]
```diff
@@ -1,26 +1,80 @@
...
+	Version       int32       `json:"version,omitempty"`
 }
 
 func NewFormControlRef() *FormControlRef {
 	return &FormControlRef{
 		Ref: NewRef(FormControlRefTable),
 	}
 }
+
+func (o *FormControlRef) ValidateCreateRequireFields(formControlName string, ref string, value string) error {
```
Comment (starkmichelsc) [inline pkg/domain/form_control_ref.go]: refactor to create a single func that takes in a string and an errMessage. this new func could be re-used more readily for other fields

**Review by starkmichelsc — CHANGES_REQUESTED** [review]

File: pkg/domain/form_control.go [hunk 8]
```diff
@@ -10,13 +10,27 @@ type FormControlList struct {
 }
 
 type FormControl struct {
-	Id          int32  `json:"id,omitempty"`
-	Ref         string `json:"ref,omitempty" validate:"required"`
+	Id          string `json:"id,omitempty"`
```
Comment (starkleem) [inline pkg/domain/form_control.go]: Should be an int not a string

**Review by starkleem — DISMISSED** [review]

File: pkg/starkapi/forms_api.go (hunk 6 above)
Comment (starksweeneym) [reply pkg/starkapi/forms_api.go:92]: Created new issue for this and will create a new PR

https://controlfreak.atlassian.net/browse/TSP-5607

File: pkg/domain/form_control_ref.go (hunk 2 above)
Comment (starkmichelsc) [reply pkg/domain/form_control_ref.go]: please add a meaningful description to the logger.Error. the actual err should be returned

File: pkg/domain/form_control_ref.go [hunk 9]
```diff
@@ -1,26 +1,75 @@
...
+
+func (o *FormControlRef) BuildFormControlRefForCreate(formControl FormControl, ref string, value string) error {
+	err := formControl.Validate()
+	if err != nil {
+		logger.Error(err)
+		return errors.New(fmt.Sprintf(errInvalidFormControl, formControl.Name))
+	}
+	err = json.Unmarshal([]byte(formControl.Control), &formControl.ControlJSON)
+	if err != nil {
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/domain/form_control_ref.go]: please add a meaningful description to the logger.Error

File: pkg/starkapi/forms_api.go [hunk 10]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+	control, err = formsApi.GetControlByName(formControlName)
 	if err != nil {
-		return control, err
+		logger.Error(err)
+		return controlRef, err
+	}
+
+	if control.Id < 1 || control.Name != formControlName {
+		newError := errors.New(fmt.Sprintf(errInvalidControlNameProvided, formControlName))
+		logger.Error(newError)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go:122]: please add a meaningful description to the logger.Error

File: pkg/starkapi/forms_api.go [hunk 11]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, newError
+	}
+
+	err = controlRef.BuildFormControlRefForCreate(control, ref, value)
+	if err != nil {
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful description to the logger.Error

File: pkg/starkapi/forms_api.go [hunk 12]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
+	}
+
+	url := fmt.Sprintf("%s/%s%s", formsApi.baseUrl(), ref, formsApi.controlsPrefix())
+
+	body, err := json.Marshal(controlRef)
+	if err != nil {
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful description to the logger.Error

File: pkg/starkapi/forms_api.go [hunk 13]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
 	}
 
 	resp, err := formsApi.client.post(url, body)
 	if err != nil {
-		return control, err
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful description to the logger.Error

File: pkg/starkapi/forms_api.go [hunk 14]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
 	}
 
 	if len(resp) > 0 {
-		err = json.Unmarshal(resp, &control)
+		err = json.Unmarshal(resp, &controlRef)
 		if err != nil {
-			return control, err
+			logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful description to the logger.Error

**Review by starkmichelsc — CHANGES_REQUESTED** [review]

**Review by starkshears — DISMISSED** [review]

File: pkg/starkapi/forms_api.go (hunk 10 above)
Comment (starksweeneym) [reply pkg/starkapi/forms_api.go:122]: I believe this already has a meaningful error. "no form control found with name provided : [%s], formControlName". If this requires more information then re-request changes please. Thanks

File: pkg/starkapi/forms_api.go [hunk 15]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
-func (formsApi *FormsApi) CreateControlOnRef(control domain.FormControl) (domain.FormControl, error) {
-	err := control.Validate()
+func (formsApi *FormsApi) CreateControlOnRef(formControlName string, ref string, value string) (domain.FormControlRef, error) {
+	var control domain.FormControl
+	var controlRef domain.FormControlRef
+
+	err := controlRef.ValidateStringParams(formControlName, errNoControlNameProvided)
 	if err != nil {
-		return control, err
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful log message

File: pkg/starkapi/forms_api.go [hunk 16]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
+	}
+
+	err = controlRef.ValidateStringParams(ref, errNoRefProvided)
+	if err != nil {
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful log message

File: pkg/starkapi/forms_api.go [hunk 17]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
 	}
 
-	url := fmt.Sprintf("%s/%s%s", formsApi.baseUrl(), control.Ref, formsApi.controlsPrefix())
+	err = controlRef.ValidateStringParams(value, errNoValueProvided)
+	if err != nil {
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful log message

File: pkg/starkapi/forms_api.go [hunk 18]
```diff
@@ -80,30 +89,67 @@ func (formsApi *FormsApi) GetControlByName(name string) (domain.FormControl, err
...
+		return controlRef, err
+	}
 
-	body, err := json.Marshal(control)
+	control, err = formsApi.GetControlByName(formControlName)
 	if err != nil {
-		return control, err
+		logger.Error(err)
```
Comment (starkmichelsc) [inline pkg/starkapi/forms_api.go]: please add a meaningful log message

**Review by starkmichelsc — CHANGES_REQUESTED** [review]

**Review by starkmichelsc — APPROVED** [review]

**Review by starkballc — APPROVED** [review]

**Review by starkshears — APPROVED** [review]

**Pull request merged**